storage/src/system/auction/detail.rs

```diff
@@ -50,20 +55,165 @@ where
     provider.write(uref, value)
 }
 
-pub fn get_validator_bids<P>(provider: &mut P) -> Result<ValidatorBids, Error>
+#[derive(Debug, Default)]
+pub struct ValidatorBidsDetail {
+    validator_bids: ValidatorBids,
+    validator_credits: ValidatorCredits,
+}
+
+impl ValidatorBidsDetail {
+    pub fn new() -> Self {
+        ValidatorBidsDetail {
+            validator_bids: BTreeMap::new(),
+            validator_credits: BTreeMap::new(),
+        }
+    }
+
+    /// Inserts a validator bid.
+    pub fn insert_bid(
+        &mut self,
+        validator: PublicKey,
+        validator_bid: Box<ValidatorBid>,
+    ) -> Option<Box<ValidatorBid>> {
+        self.validator_bids.insert(validator, validator_bid)
+    }
+
+    /// Inserts a validator credit.
+    pub fn insert_credit(
+        &mut self,
+        validator: PublicKey,
+        era_id: EraId,
+        validator_credit: Box<ValidatorCredit>,
+    ) {
+        let credits = &mut self.validator_credits;
+
+        match credits.entry(validator.clone()) {
+            Entry::Vacant(vacancy) => {
+                let mut inner = BTreeMap::new();
+                inner.insert(era_id, validator_credit);
+                vacancy.insert(inner);
+            }
+            Entry::Occupied(outer) => {
+                let mut inner = outer.remove();
+                match inner.entry(era_id) {
+                    Entry::Vacant(vacancy) => {
+                        vacancy.insert(validator_credit);
```

Comment:
I think we may have a bug here. Don't we want to put the entry back here?
Or do we maybe want this function to be something simplified like this?:
```
    credits.entry(validator_id).and_modify(|inner| {
        inner.entry(era_id).and_modify(|_| warn!(?validator, ?era_id, "multiple validator credit entries in same era")).or_insert(validator_credit);
    }).or_insert_with(|| {
        let mut inner = BTreeMap::new();
        inner.insert(era_id, validator_credit);
        inner
    });
```